design/aws-ecs-service-discovery-integration.md

```diff
@@ -0,0 +1,235 @@
+# AWS ECS - Support for AWS Cloud Map (Service Discovery) Integration
+
+To address issue [#1554](https://github.com/awslabs/aws-cdk/issues/1554), the
+ECS construct library should provide a way to set
+[`serviceRegistries`](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_CreateService.html#ECS-CreateService-request-serviceRegistries)
+on their ECS service at the L1 level.
+
+
+## General approach
+Rather than having the customer instantiate Cloud Map constructs directly and
+thus be forced to learn the API of a separate construct library, we should
+allow the customer to pass in the domain name and other configuration minimally
+needed to instantiate a Cloud Map namespace within an ECS cluster and create a
+Cloud Map service for each ECS service.
+
+The current proposal is to add a method on the ECS Cluster construct,
+`addNamespace` that would take a set of properties that include the namespace
+type (Public DNS, Private DNS or Http) and namespace name (domain name).
+
+While it is possible to use more than one namespace for services in an ECS
+cluster, realistically, we would expect ECS customers to only have one
+namespace per given cluster. A Cloud Map service within that namespace would
+then be created for each ECS service using service discovery in the cluster,
+and would be discoverable by service name within that namespace, e.g.
+frontend.mydomain.com, backend.mydomain.com, etc.
+
+ECS will automatically register service discovery instances that are accessible
+by IP address (IPv4 only) on the createService API call.
+
+// FIXME public namespace needs to be imported?
+
+## Code changes
+
+The following are the new methods/interfaces that would be required for the proposed approach:
+
+#### Cluster#addNamespace(options: NamespaceOptions)
+
+This will allow adding a Cloud Map namespace, which will be accessible as a
+member on the cluster. In the case of a Private DNS Namespace, a Route53 hosted
+zone will be created for the customer.
+
+```ts
+export interface NamespaceOptions {
+  /**
+   * The domain name for the namespace, such as foo.com
+   */
+  name: string;
+
+  /**
+   * The type of CloudMap Namespace to create in your cluster
+   *
+   * @default PrivateDns
+   */
+  type?: cloudmap.NamespaceType;
+
+  /**
+   * The Amazon VPC that you want to associate the namespace with. Required for Private DNS namespaces
+   *
+   * @default VPC of the cluster for Private DNS Namespace, otherwise none
+   */
+  vpc?: ec2.IVpcNetwork;
+}
+```
+
+#### service#enableServiceDiscovery(options: ServiceDiscoveryOptions)
```

Comment:
Why not have this as constructor parameters?

And I feel if you enable a namespace on the cluster, service discovery should be enabled by default, no?

Reply:
I believe that not all services within a cluster will necessarily want to be hooked into Cloud Map, though would love input from @nathanpeck or @herrhound on this. 

I think I leaned this way bc I liked thinking of service discovery options as its own bundle, and enabling it as a modular operation, but from the point of view of the CDK you're right in that it doesn't matter. Implementation of this is [here](https://github.com/awslabs/aws-cdk/pull/2065/files#diff-7f1d40bd7ae949a75f338575b348b20fR197) and [here](https://github.com/awslabs/aws-cdk/pull/2065/files#diff-abac9fea95aeaefb9d27306c90dac45cR75) so let me know if you think that all can just live in the constructor.

Reply:
I was thinking the constructor would have something like this:

```ts
/** @default true */
registerInCloudMap?: boolean;

cloudMapServiceOptions?: CloudMapServiceOptions;
```

Also we're going to have to make a decision on whether to call it "ServiceDiscovery" or "CloudMap". Right now we're doing both, which seems confusing.

Reply:
Yes; it is already confusing in our docs since we mention service discovery but a lot of those pages redirect to AWS Cloud Map. Thoughts from @herrhound?